fix(metadata): massage sourceUrls starting with git@

scarrupt opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox are checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes
<!-- Describe what behavior is changed by this PR. -->
`massageUrl` function now returns sourceURLs starting with git@ in HTTPS format.

## Context

<!-- Describe why you're making these changes if it's not already explained in a related issue. -->
<!-- If you're closing an existing problem with this pull request, use the keyword Closes #issue_number -->
Closes #18015

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after creating your PR, as doing so causes us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice reviewed (commented):
this will probably break self-hosted users who are using ssh

rarkins commented:
> this will probably break self-hosted users who are using ssh

So should it be fixed, or just documented and part of v33?

viceice commented:
don't know how that can be fixed. 🤷‍♂️

also don't know how many users we break with this.

scarrupt commented:
What do you meaning by breaking? Is this sourceUrl used to checking out code?

viceice commented:
> What do you meaning by breaking? Is this sourceUrl used to checking out code?

had another look. no it's not used for git interaction.

but simply replacing ssh with https won't work for a lot of platforms, as there are many where the https and ssh urls are totally different.

scarrupt commented:
> > What do you meaning by breaking? Is this sourceUrl used to checking out code?
> 
> had another look. no it's not used for git interaction.
> 
> but simply replacing ssh with https won't work for a lot of platforms, as there are many where the https and ssh urls are totally different.

Is there any code flow not going to `addMetadata` function because currently, Renovate is also going to replace sourceURL starting with 'ssh' by 'HTTPS'?

https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/metadata.ts#L144

scarrupt commented:
> > > What do you meaning by breaking? Is this sourceUrl used to checking out code?
> > 
> > 
> > had another look. no it's not used for git interaction.
> > but simply replacing ssh with https won't work for a lot of platforms, as there are many where the https and ssh urls are totally different.
> 
> Is there any code flow not going to `addMetadata` function because currently, Renovate is also going to replace sourceURL starting with 'ssh' by 'HTTPS'?
> 
> https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/metadata.ts#L144

My bad, it doesn't. So would it mean only ssh should not be replaced?

scarrupt commented:
@viceice @rarkins I have modified `massageUrl` function so it can extract the hostname from sourceUrls starting with git@ and return them in HTTPS format, in a similar fashion to `massageGithubUrl`. Any reason that the function doesn't massage GitLab URLs starting with git@ prefix?

rarkins commented:
No reason I'm aware of

viceice commented:
probably nobody raised a request for gitlab yet. 🤗

scarrupt commented:
> probably nobody raised a request for gitlab yet. 🤗

If it makes sense, I can add it to `massageGitlabUrl`

scarrupt commented:
> > probably nobody raised a request for gitlab yet. 🤗
> 
> If it makes sense, I can add it to `massageGitlabUrl`

@viceice @rarkins would you like I massage @git for Gitlab or not?

scarrupt commented:
@viceice @rarkins I added massaging of the git@ URL for Gitlab as well. I extracted the code to replace git@ with HTTPS in a function. I finished developing the fix. Let me know if I need to change some code.

viceice commented on lib/modules/datasource/metadata.spec.ts:
```diff
@@ -231,6 +232,50 @@ describe('modules/datasource/metadata', () => {
+      ${'ssh://git@github.com/user/repo'}
+      ${'git://github.com/user/repo'}
+      ${'https://www.github.com/user/repo'}
+      ${'https://user.github.com/repo'}
```
is this a valid url?

rarkins replied:
Doesn't seem to be

scarrupt replied:
I added that test case for the GitHub pages branch although those URLs are not valid anymore:

https://github.blog/changelog/2021-01-29-github-pages-will-stop-redirecting-pages-sites-from-github-com-after-april-15-2021/

Yet, I’m not sure if it should be fixed in this issue or create another issue because it is going out of scope

viceice replied:
ok, leave it for now.

viceice commented on lib/modules/datasource/metadata.spec.ts:
```diff
@@ -231,6 +232,50 @@ describe('modules/datasource/metadata', () => {
+      ${'https://gitlab.com/user/repo/tree/master'}
+      ${'http://gitlab.com/user/repo/'}
+      ${'http://gitlab.com/user/repo.git'}
+      ${'git@gitlab.com:user/repo.git'}
```
what about gitlab ssh urls? we should do same as for github

scarrupt replied:
Same comment as above about the scope of the issue. I'm happy to add those URLs in another PR/issue.

viceice commented on lib/modules/datasource/metadata.ts:
```diff
@@ -13,22 +13,26 @@ const githubPages = regEx('^https://([^.]+).github.com/([^/]+)$');
     return '';
   }
-  if (parsedUrl.hostname.includes('gitlab')) {
+  if (hostname.includes('gitlab')) {
```
we should probably use our util platform detection code here instead 🤔

scarrupt replied:
@viceice I don't think I'm able to use platform detection here because the current code is massaging URLs that are not from GitHub or GitLab. If an URL is neither one of those platforms, Renovate massages it as a GitHub URL.

scarrupt replied:
Renovate won't massage URLs that are not valid. The platform detection method will return null for invalid URLs and URLs not from GitLab or Github.

scarrupt replied:
The platform detection method doesn't provide a way to differentiate between valid and invalid URLs.

viceice replied:
i know, but it uses hostrules to detect platform for arbitrary urls, so we can match more urls (for self-hosted github / gitlab) too

https://github.com/renovatebot/renovate/blob/6ecd1fdc68c47a794e768a5d37fb37380e32d711/lib/util/common.ts#L14

scarrupt replied:
I see your point, and if I use `detectPlatform` function there is a change in behavior with URLs that are not from GitLab or GitHub (e.g example.com). Renovate will delete sourceUrl. Is it something that Renovate should do?

scarrupt replied:
For example, the existing test `Should handle failed parsing of sourceUrls for other` will fail because sourceUrl is deleted:

https://github.com/scarrupt/renovate/blob/f7394b04ba878d43686763872adbf2a5ed196758/lib/modules/datasource/metadata.spec.ts#L142

scarrupt commented:
I added a new commit because I noticed that I was not checking against the correct massaging function (massageUrl) for `Should return an empty string when massaging an invalid url` test.

scarrupt commented:
@rarkins @viceice 

Will you create issues for GitHub pages and GitLab ssh sources, or should I? I'll create new PRs

What is your decision about detectPlatform? As I said, I don't think metadata can use `detectPlatform` in its current state because it'll delete sourceURLs that are valid. Another example that I found in the code is bitbucket URLs for npm:

https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/npm/get.ts#L22-L24

rarkins commented:
> Will you create issues for GitHub pages and GitLab ssh sources, or should I? I'll create new PRs

Please create issues

> What is your decision about detectPlatform? As I said, I don't think metadata can use `detectPlatform` in its current state because it'll delete sourceURLs that are valid. 

Could you explain more on this point?

scarrupt commented:
> > Will you create issues for GitHub pages and GitLab ssh sources, or should I? I'll create new PRs
> 
> Please create issues
> 
> > What is your decision about detectPlatform? As I said, I don't think metadata can use `detectPlatform` in its current state because it'll delete sourceURLs that are valid.
> 
> Could you explain more on this point?

@rarkins Sure. Let's compare the previous/current implementation to an implementation with `detectPlatform` by using some examples:

**Previous/current implementation**
sourceURL=https://github.com/user/repo -> massagedSourceURL=https://github.com/user/repo
sourceURL=https://gitlab.com/user/repo -> massagedSourceURL=https://gitlab.com/user/repo
sourceURL=https:/bitbucket.com/user/repo -> massagedSourceURL=https://bitbucket.com/user/repo
sourceURL=invalidURL -> null

**Implementation with detectPlatform**
sourceURL=https://github.com/user/repo -> massagedSourceURL=https://github.com/user/repo
sourceURL=https://gitlab.com/user/repo -> massagedSourceURL=https://gitlab.com/user/repo
sourceURL=https:/bitbucket.com/user/repo -> massagedSourceURL=null
sourceURL=invalidURL -> null

You can notice that detectPlatform cannot distinguish between an invalid URL or an URL that is not a GitHub or a GitLab URL. As a result, Renovate will remove sourceURL, which I think is a regression from the current implementation. Does it make sense, or am I missing something?

rarkins commented:
Maybe there's a misunderstanding. What I mean is to replace lines like this:

```
if (hostname.includes('gitlab')) {
```

with:

```
if (detectPlatform(hostname) === 'gitlab') {
```

This itself does not delete any source URLs

scarrupt commented:
> Maybe there's a misunderstanding. What I mean is to replace lines like this:
> 
> ```
> if (hostname.includes('gitlab')) {
> ```
> 
> with:
> 
> ```
> if (detectPlatform(hostname) === 'gitlab') {
> ```
> 
> This itself does not delete any source URLs

@rarkins. I see. Sorry, I misunderstood your comment. I'll make the change and open issues.

scarrupt commented:
> > Maybe there's a misunderstanding. What I mean is to replace lines like this:
> > ```
> > if (hostname.includes('gitlab')) {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > with:
> > ```
> > if (detectPlatform(hostname) === 'gitlab') {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This itself does not delete any source URLs
> 
> @rarkins. I see. Sorry, I misunderstood your comment. I'll make the change and open issues.

@rarkins I made the change.
